File: references/betterlectio/ios/BetterLectio/AssignmentParser.swift
//
//  AssignmentParser.swift
//  BetterLectio
//

import Foundation
import SwiftSoup

/// Parser for assignments and exercises from Lectio.
enum AssignmentParser {
    
    // MARK: - Parse Assignments

    /// Parses assignments table from OpgaverElev.aspx HTML
    static func parseAssignments(from html: String) throws -> [Assignment] {
        let doc = try SwiftSoup.parse(html)

        guard let table = try doc.select("table#s_m_Content_Content_ExerciseGV").first() else {
            print("⚠️ Assignments table not found")
            return []
        }

        let rows = try table.select("tbody tr")
        var assignments: [Assignment] = []

        // Danish date formatter for deadline parsing: "3/12-2025 14:00"
        let deadlineFormatter = DateFormatter()
        deadlineFormatter.dateFormat = "d/M-yyyy HH:mm"
        deadlineFormatter.locale = Locale(identifier: "da_DK")

        for row in rows {
            // Skip header row (has th elements)
            let headers = try row.select("th")
            if !headers.isEmpty() { continue }

            // Get desktop-only cells
            let cells = try row.select("td.OnlyDesktop")
            guard cells.count >= 11 else { continue }

            // Week
            let week = try cells[0].text().trimmingCharacters(in: .whitespacesAndNewlines)

            // Hold (class) + element ID
            let holdSpan = try cells[1].select("span[data-lectiocontextcard]").first()
            let hold = try holdSpan?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? cells[1].text().trimmingCharacters(in: .whitespacesAndNewlines)
            let holdElementId = try holdSpan?.attr("data-lectiocontextcard") ?? ""

            // Title + detail URL + exercise ID
            let titleLink = try cells[2].select("a").first()
            let title = try titleLink?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? cells[2].text().trimmingCharacters(in: .whitespacesAndNewlines)
            let detailURL = try titleLink?.attr("href") ?? ""

            // Extract exerciseid from URL
            var exerciseId = ""
            if let range = detailURL.range(of: "exerciseid=(\\d+)", options: .regularExpression) {
                exerciseId = String(detailURL[range]).replacingOccurrences(of: "exerciseid=", with: "")
            }

            guard !exerciseId.isEmpty else { continue }

            // Deadline
            let deadline = try cells[3].text().trimmingCharacters(in: .whitespacesAndNewlines)
            let deadlineDate = deadlineFormatter.date(from: deadline)

            // Student time
            let studentTime = try cells[4].text().trimmingCharacters(in: .whitespacesAndNewlines)

            // Status
            let statusText = try cells[5].text().trimmingCharacters(in: .whitespacesAndNewlines)
            let hasWaitingSpan = try !cells[5].select("span.exercisewait").isEmpty()
            let status: AssignmentStatus
            if hasWaitingSpan || statusText == "Venter" {
                status = .waiting
            } else if statusText == "Afleveret" {
                status = .submitted
            } else if statusText.contains("Ikke afleveret") {
                status = .notSubmitted
            } else if statusText == "Mangler" {
                status = .missing
            } else {
                status = .waiting
            }

            // Absence
            let absence = try cells[6].text().trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty

            // Awaiting
            let awaiting = try cells[7].text().trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty

            // Assignment note
            let assignmentNote = try cells[8].text().trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty

            // Grade (may contain grade + grade note separated by <br>)
            let gradeCell = cells[9]
            let gradeHTML = try gradeCell.html()
            let gradeParts = gradeHTML.components(separatedBy: "<br>")
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
            let grade = gradeParts.first
            let gradeNote = gradeParts.dropFirst().first

            // Student note
            let studentNote = try cells[10].text().trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty

            assignments.append(Assignment(
                id: exerciseId,
                week: week,
                hold: hold,
                holdElementId: holdElementId,
                title: title,
                deadline: deadline,
                deadlineDate: deadlineDate,
                studentTime: studentTime,
                status: status,
                absence: absence,
                awaiting: awaiting,
                assignmentNote: assignmentNote,
                grade: grade,
                gradeNote: gradeNote,
                studentNote: studentNote,
                detailURL: detailURL
            ))
        }

        print("📝 Parsed \(assignments.count) assignments")
        return assignments
    }

    // MARK: - Parse Assignment Detail

    /// Parses the assignment detail page (ElevAflevering.aspx) into structured content.
    static func parseAssignmentDetail(from html: String) throws -> AssignmentDetail {
        let doc = try SwiftSoup.parse(html)

        // --- Section 1: Opgaveoplysninger ---
        guard let infoSection = try doc.select("#m_Content_registerAfl_pa").first() else {
            throw LectioError.parsingError("Assignment info section not found")
        }

        // Title
        let title = try doc.select("#m_Content_NameLbl").first()?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        // Description files (Opgavebeskrivelse)
        var descriptionFiles: [AssignmentFile] = []
        let descLinks = try infoSection.select("a[id*=showdocumentHyperlnk]")
        for link in descLinks {
            let href = try link.attr("href")
            let text = try link.text().trimmingCharacters(in: .whitespacesAndNewlines)
            guard !href.isEmpty else { continue }
            descriptionFiles.append(AssignmentFile(id: href, name: text, url: href))
        }

        // Parse key-value rows from the info table
        let infoRows = try infoSection.select("table.ls-std-table-inputlist tr")
        var hold = ""
        var gradeScale: String?
        var teacher = ""
        var studentTime = ""
        var deadline = ""
        var assignmentNote: String?
        var inDescription = false

        for row in infoRows {
            let th = try row.select("th").first()?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            let td = try row.select("td").first()

            if th.hasPrefix("Hold") {
                hold = try td?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            } else if th.hasPrefix("Karakterskala") {
                gradeScale = try td?.text().trimmingCharacters(in: .whitespacesAndNewlines)
            } else if th.hasPrefix("Ansvarlig") {
                teacher = try td?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            } else if th.hasPrefix("Elevtid") {
                studentTime = try td?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            } else if th.hasPrefix("Afleveringsfrist") {
                deadline = try td?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            } else if th.hasPrefix("Opgavenote") {
                let noteText = try td?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                assignmentNote = noteText.isEmpty ? nil : noteText
            } else if th.contains("undervisningsbeskrivelse") {
                let val = try td?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                inDescription = val == "Ja"
            }
        }

        // --- Section 2: Afleveres af (student status) ---
        var awaiting: String?
        var status: String?
        var completed = false
        var grade: String?
        var gradeNote: String?
        var studentNote: String?

        if let studentTable = try doc.select("#m_Content_StudentGV").first() {
            let studentRows = try studentTable.select("tr")
            // Skip header row (first tr with th elements)
            for row in studentRows {
                let cells = try row.select("td")
                guard cells.count >= 7 else { continue }

                // cells: [0]=photo, [1]=name, [2]=awaiting, [3]=status, [4]=completed, [5]=grade, [6]=gradeNote, [7]=studentNote
                awaiting = try cells[2].text().trimmingCharacters(in: .whitespacesAndNewlines)
                if awaiting?.isEmpty == true { awaiting = nil }

                status = try cells[3].text().trimmingCharacters(in: .whitespacesAndNewlines)
                if status?.isEmpty == true { status = nil }

                // Completed checkbox
                let checkbox = try cells[4].select("input[type=checkbox]").first()
                if let cb = checkbox {
                    completed = try cb.hasAttr("checked")
                }

                let gradeText = try cells[5].text().trimmingCharacters(in: .whitespacesAndNewlines)
                grade = gradeText.isEmpty ? nil : gradeText

                let gradeNoteText = try cells[6].text().trimmingCharacters(in: .whitespacesAndNewlines)
                gradeNote = gradeNoteText.isEmpty ? nil : gradeNoteText

                if cells.count > 7 {
                    let studentNoteText = try cells[7].text().trimmingCharacters(in: .whitespacesAndNewlines)
                    studentNote = studentNoteText.isEmpty ? nil : studentNoteText
                }

                break // Only need the first data row (the student)
            }
        }

        // --- Section 3: Opgaveindlæg (submissions) ---
        var submissions: [AssignmentSubmission] = []

        if let submissionTable = try doc.select("#m_Content_RecipientGV").first() {
            // Check for empty state
            let noRecord = try submissionTable.select("span.norecord")
            if noRecord.isEmpty() {
                let subRows = try submissionTable.select("tr")
                var index = 0
                for row in subRows {
                    let cells = try row.select("td")
                    guard cells.count >= 4 else { continue }

                    let timestamp = try cells[0].text().trimmingCharacters(in: .whitespacesAndNewlines)
                    let user = try cells[1].text().trimmingCharacters(in: .whitespacesAndNewlines)
                    let commentText = try cells[2].text().trimmingCharacters(in: .whitespacesAndNewlines)
                    let comment = commentText.isEmpty ? nil : commentText

                    // Document link
                    var document: AssignmentFile?
                    if let docLink = try cells[3].select("a").first() {
                        let href = try docLink.attr("href")
                        let docName = try docLink.text().trimmingCharacters(in: .whitespacesAndNewlines)
                        if !href.isEmpty {
                            document = AssignmentFile(id: href, name: docName, url: href)
                        }
                    }

                    submissions.append(AssignmentSubmission(
                        id: "\(index)",
                        timestamp: timestamp,
                        user: user,
                        comment: comment,
                        document: document
                    ))
                    index += 1
                }
            }
        }

        print("📋 Parsed assignment detail: \(title), \(submissions.count) submissions, \(descriptionFiles.count) description files")

        return AssignmentDetail(
            title: title,
            hold: hold,
            gradeScale: gradeScale,
            teacher: teacher,
            studentTime: studentTime,
            deadline: deadline,
            assignmentNote: assignmentNote,
            inDescription: inDescription,
            descriptionFiles: descriptionFiles,
            submissions: submissions,
            awaiting: awaiting,
            status: status,
            completed: completed,
            grade: grade,
            gradeNote: gradeNote,
            studentNote: studentNote
        )
    }
}

File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/assignments/AssignmentParser.kt
package dk.betterlectio.android.feature.assignments

import dk.betterlectio.android.core.lectio.scrape.AspNetForm
import dk.betterlectio.android.core.util.LectioDateUtils
import org.jsoup.Jsoup
import org.jsoup.nodes.Element

/**
 * Assignments list + detail.
 * List: iOS [AssignmentParser] (`ExerciseGV` + OnlyDesktop) with Flutter column layout.
 * Detail: iOS label-based info table + RecipientGV + StudentGV.
 */
object AssignmentParser {
    fun parseList(html: String): List<AssignmentItem> {
        val doc = Jsoup.parse(html)
        val table = doc.getElementById("s_m_Content_Content_ExerciseGV")
            ?: doc.selectFirst("table[id*=ExerciseGV]")
            ?: doc.selectFirst("table")
            ?: return emptyList()
        return table.select("tr").mapNotNull { row ->
            if (row.select("th").isNotEmpty()) return@mapNotNull null
            parseRow(row)
        }
    }

    private fun parseRow(row: Element): AssignmentItem? {
        val cols = desktopCells(row)
        if (cols.size < 6) return null
        val week = cols[0].text().trim().toIntOrNull() ?: 0
        val holdSpan = cols[1].selectFirst("span[data-lectiocontextcard]")
        val team = holdSpan?.text()?.trim() ?: cols[1].text().trim()
        val holdElementId = holdSpan?.attr("data-lectiocontextcard")?.ifBlank { null }
        val link = cols[2].selectFirst("a") ?: return null
        val href = link.attr("href")
        val id = AspNetForm.queriesFromUrl(href)["exerciseid"]
            ?: Regex("""exerciseid=(\d+)""").find(href)?.groupValues?.get(1)
            ?: return null
        val title = link.text().trim()
        val deadline = LectioDateUtils.parseLectioDate(cols[3].text().trim())
        val studentTime = cols.getOrNull(4)?.text()?.replace(',', '.')?.trim()?.toDoubleOrNull() ?: 0.0
        val statusCell = cols.getOrNull(5)
        val statusText = statusCell?.text()?.trim().orEmpty()
        val status = when {
            statusCell?.selectFirst("span.exercisewait") != null -> "Venter"
            statusText.equals("Venter", true) -> "Venter"
            statusText.equals("Afventer", true) -> "Venter"
            else -> statusText
        }
        val absence = cols.getOrNull(6)?.text()?.trim().orEmpty()
        val awaits = cols.getOrNull(7)?.text()?.trim().orEmpty()
        val note = cols.getOrNull(8)?.text()?.trim().orEmpty()
        var grade: String? = null
        var gradeNote: String? = null
        cols.getOrNull(9)?.let { gradeCell ->
            val parts = gradeCell.html()
                .split(Regex("""<br\s*/?>""", RegexOption.IGNORE_CASE))
                .map { Jsoup.parse(it).text().trim() }
                .filter { it.isNotEmpty() }
            grade = parts.getOrNull(0)
            gradeNote = parts.getOrNull(1)
        }
        val studentNote = cols.getOrNull(10)?.text()?.trim()?.ifBlank { null }
        return AssignmentItem(
            id = id,
            title = title,
            team = team,
            week = week,
            deadline = deadline,
            status = status,
            studentTime = studentTime,
            awaits = awaits,
            note = note,
            absence = absence,
            grade = grade,
            gradeNote = gradeNote,
            studentNote = studentNote,
            holdElementId = holdElementId,
            detailUrl = href.ifBlank { null },
        )
    }

    fun parseDetail(html: String, item: AssignmentItem): AssignmentDetail {
        val doc = Jsoup.parse(html)
        val title = doc.getElementById("m_Content_NameLbl")?.text()?.trim().orEmpty().ifBlank { item.title }

        val infoSection = doc.selectFirst("#m_Content_registerAfl_pa")
        var hold = item.team
        var grading = doc.getElementById("m_Content_gradeScaleIdLbl")?.text()?.trim().orEmpty()
        var responsible = ""
        var studentTime = item.studentTime
        var description = item.note
        val files = mutableListOf<Pair<String, String>>()

        if (infoSection != null) {
            // Description files — iOS: a[id*=showdocumentHyperlnk]
            infoSection.select("a[id*=showdocumentHyperlnk], #m_Content_ExerciseFilePnl a").forEach { a ->
                val href = a.attr("href").trim()
                if (href.isEmpty()) return@forEach
                val name = a.text().trim().ifBlank { "Fil" }
                val url = absolutize(href)
                files += name to url
            }

            for (row in infoSection.select("table.ls-std-table-inputlist tr, table tr")) {
                val th = row.selectFirst("th")?.text()?.trim().orEmpty()
                val td = row.selectFirst("td") ?: continue
                when {
                    th.startsWith("Hold") -> hold = td.text().trim().ifBlank { hold }
                    th.startsWith("Karakterskala") -> grading = td.text().trim()
                    th.startsWith("Ansvarlig") -> responsible = td.text().trim()
                    th.startsWith("Elevtid") -> {
                        val raw = td.text().trim().ifBlank {
                            doc.getElementById("m_Content_WeightLbl")?.text()?.trim().orEmpty()
                        }
                        studentTime = raw.split(Regex("""\s+""")).firstOrNull()
                            ?.replace(',', '.')
                            ?.toDoubleOrNull()
                            ?: studentTime
                    }
                    th.startsWith("Opgavenote") -> description = td.text().trim().ifBlank { description }
                    th.contains("undervisningsbeskrivelse") -> { /* optional flag */ }
                }
            }
            // Fallback WeightLbl if Elevtid row missing
            if (studentTime == item.studentTime) {
                val weight = doc.getElementById("m_Content_WeightLbl")?.text()?.trim().orEmpty()
                studentTime = weight.split(Regex("""\s+""")).firstOrNull()
                    ?.replace(',', '.')
                    ?.toDoubleOrNull()
                    ?: studentTime
            }
        } else {
            // Legacy fallbacks (old Android paths)
            description = doc.selectFirst("#m_Content_DescriptionPnl, .exercise-description")
                ?.text()?.trim().orEmpty().ifBlank { item.note }
            doc.select("#m_Content_ExerciseFilePnl a").forEach { a ->
                val href = a.attr("href")
                if (href.isNotBlank()) files += a.text().trim().ifBlank { "Fil" } to absolutize(href)
            }
            responsible = doc.selectFirst("[data-lectiocontextcard^=T]")?.text()?.trim().orEmpty()
        }

        // Student status row
        var awaits = item.awaits
        var status = item.status
        var completed = false
        var studentGrade: String? = item.grade
        var studentGradeNote: String? = item.gradeNote
        doc.selectFirst("#m_Content_StudentGV")?.select("tr")?.forEach { row ->
            val cells = row.select("td")
            if (cells.size < 6) return@forEach
            awaits = cells.getOrNull(2)?.text()?.trim()?.ifBlank { awaits } ?: awaits
            status = cells.getOrNull(3)?.text()?.trim()?.ifBlank { status } ?: status
            completed = cells.getOrNull(4)?.selectFirst("input[type=checkbox][checked]") != null ||
                cells.getOrNull(4)?.selectFirst("input[type=checkbox]")?.hasAttr("checked") == true
            studentGrade = cells.getOrNull(5)?.text()?.trim()?.ifBlank { null } ?: studentGrade
            studentGradeNote = cells.getOrNull(6)?.text()?.trim()?.ifBlank { null } ?: studentGradeNote
            return@forEach // first data row only
        }

        // Submissions
        val submissions = mutableListOf<AssignmentSubmission>()
        val recipient = doc.selectFirst("#m_Content_RecipientGV")
        if (recipient != null && recipient.select("span.norecord").isEmpty()) {
            var index = 0
            for (row in recipient.select("tr")) {
                val cells = row.select("td")
                if (cells.size < 4) continue
                val timestamp = cells[0].text().trim()
                val user = cells[1].text().trim()
                if (timestamp.isBlank() && user.isBlank()) continue
                val comment = cells.getOrNull(2)?.text()?.trim()?.ifBlank { null }
                val docLink = cells.getOrNull(3)?.selectFirst("a")
                submissions += AssignmentSubmission(
                    id = "$index",
                    timestamp = timestamp,
                    user = user,
                    comment = comment,
                    documentName = docLink?.text()?.trim()?.ifBlank { null },
                    documentUrl = docLink?.attr("href")?.takeIf { it.isNotBlank() }?.let { absolutize(it) },
                )
                index++
            }
        }

        return AssignmentDetail(
            item = item.copy(
                title = title,
                team = hold,
                status = status,
                awaits = awaits,
                studentTime = studentTime,
                note = description.ifBlank { item.note },
                grade = studentGrade,
                gradeNote = studentGradeNote,
            ),
            description = description,
            files = files.distinctBy { it.second },
            responsible = responsible,
            grading = grading,
            submissions = submissions,
            completed = completed,
            studentGrade = studentGrade,
            studentGradeNote = studentGradeNote,
        )
    }

    private fun desktopCells(row: Element): List<Element> {
        val desktop = row.select("td.OnlyDesktop")
        if (desktop.isNotEmpty()) return desktop
        val nonMobile = row.select("td:not(.OnlyMobile)")
        if (nonMobile.isNotEmpty()) return nonMobile
        return row.select("td")
    }

    private fun absolutize(href: String): String = when {
        href.startsWith("http") -> href
        href.startsWith("/") -> "https://www.lectio.dk$href"
        else -> "https://www.lectio.dk/$href"
    }
}
